storage/paths/db_index.go

```diff
@@ -0,0 +1,986 @@
+			if isLocked(lockMap[wft].writeTs) || isLocked(lockMap[wft].readTs) {
+				return false, xerrors.Errorf("cannot acquire writelock for sector %v filetype %d already locked", sector, wft)
+			}
+		}
+
+		// Check if we can acquire read locks
+		// Conditions: No write lock or write lock is stale
+		for _, rft := range read.AllSet() {
+			if isLocked(lockMap[rft].writeTs) {
+				return false, xerrors.Errorf("cannot acquire read lock for sector %v filetype %d already locked for writing", sector, rft)
+			}
+		}
+
+		// Acquire write locks
+		_, err = dbi.harmonyDB.Exec(ctx,
+			`UPDATE sectorlocation
+							 	SET write_ts = NOW(), write_lock_owner = $1
+							 	WHERE miner_id=$2
+							 		AND sector_num=$3
+							 		AND sector_filetype = ANY($4)`,
+			lockUuid.String(),
+			sector.Miner,
+			sector.Number,
+			write.AllSet())
+		if err != nil {
+			return false, xerrors.Errorf("acquiring write locks for sector %v fails with err: %v", sector, err)
+		}
+
+		// Acquire read locks
+		_, err = dbi.harmonyDB.Exec(ctx,
+			`UPDATE sectorlocation
+							 	SET read_ts = NOW(), read_refs = read_refs + 1
+							 	WHERE miner_id=$1
+							 		AND sector_num=$2
+							 		AND sector_filetype = ANY($3)`,
+			sector.Miner,
+			sector.Number,
+
+			read.AllSet())
+		if err != nil {
+			return false, xerrors.Errorf("acquiring read locks for sector %v fails with err: %v", sector, err)
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (dbi *DBIndex) unlock(ctx context.Context, sector abi.SectorID, read storiface.SectorFileType, write storiface.SectorFileType, lockUuid uuid.UUID) (bool, error) {
+	if read|write == 0 {
+		return false, nil
+	}
+
+	if read|write > (1<<storiface.FileTypes)-1 {
+		return false, xerrors.Errorf("unknown file types specified")
+	}
+
+	// Relinquish write locks
+	_, err := dbi.harmonyDB.Exec(ctx,
+		`UPDATE sectorlocation
+				SET write_ts = NULL, write_lock_owner = NULL
+				WHERE miner_id=$2
+					AND sector_num=$3
+					AND write_lock_owner=$4
+					AND sector_filetype = ANY($5)`,
+		lockUuid.String(),
+		sector.Miner,
+		sector.Number,
+		lockUuid.String(),
+		write.AllSet())
+	if err != nil {
+		return false, xerrors.Errorf("relinquishing write locks for sector %v fails with err: %v", sector, err)
+	}
+
+	// Relinquish read locks
+	_, err = dbi.harmonyDB.Exec(ctx,
+		`UPDATE sectorlocation
+				SET read_refs = read_refs-1,
+					read_ts = CASE WHEN read_refs - 1 = 0 THEN NULL ELSE read_ts END
+				WHERE miner_id=$1
+					AND sector_num=$2
+					AND sector_filetype = ANY($3)`,
+		sector.Miner,
+		sector.Number,
+		read.AllSet())
+	if err != nil {
+		return false, xerrors.Errorf("relinquishing read locks for sector %v fails with err: %v", sector, err)
+	}
+
+	return true, nil
+
+}
+
+func (dbi *DBIndex) StorageLock(ctx context.Context, sector abi.SectorID, read storiface.SectorFileType, write storiface.SectorFileType) error {
+
+	retries := 5
+	waitTime := 1
```

Comment:
We want to block in StorageLock until we either succeed at getting the lock, or until the context is cancelled
* There should be no retry limit
* Ideally we'd have some backoff on initial retries, starting at let's say retry after 1s, and increase to 1min, then just try every 1min

Reply:
Does an increasing backoff make sense here? Most of the time you just want to try to achieve the lock after the previous user is finished with it. If you keep on increasing the backoff, it just increases the chances of you not getting the lock the next time because a different user waiting for a shorter period will get to it first

Reply:
Also, I don't think we should try to block indefinitely. There needs to be some threshold after which you just return an API error.

For ex. if the database goes down, we don't want API callers to be stuck forever waiting for a lock their not gonna get

Reply:
But we want to block on trying to get the lock in case someone else is holding it, and the way I've read this code is that we will only attempt to get the lock for 5 seconds, which is not nearly long enough if there's a long-running operation holding the lock (which can be hours in case of PC1)

Reply:
Right. The retry parameters need to be tuned as I mentioned in a different comment but does the general concept of having a regular backoff + limit on the amount of times you retry make sense?

I could see an increasing backoff make sense when there are different task profiles with touching the same file with large variances in hold time.
I suppose you can not have a limit and rely on a timeout limit of the original RPC call (if one existst) which cancels the ctx. But I don't know if those 2 things are the case atm. Happy to reconsider if they are